### PR TITLE
Designware (DW) I2C Slave Support and Raspberry Pi Pico I2C Support

### PR DESCRIPTION
**Pull request by andrei-edward-popa**
Added designware I2C support for slave mode
Added I2C support for Raspberry Pi Pico board. I2C chip on RP2040 is DW based.

### CODE REVIEW

File: drivers/i2c/i2c_dw.c [hunk 1]
```diff
@@ -615,6 +840,14 @@ static int i2c_dw_initialize(const struct device *dev)
 	struct i2c_dw_dev_config * const dw = dev->data;
 	union ic_con_register ic_con;
 	uint32_t reg_base = get_regs(dev);
+	int ret = 0;
+
+#if defined(CONFIG_PINCTRL)
+	ret = pinctrl_apply_state(rom->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret) {
+		return ret;
+	}
```
Comment (gmarull) [inline drivers/i2c/i2c_dw.c:849]: nit: pinctrl support is independent of slave support addition, so should probably be split into another commit

File: drivers/i2c/i2c_dw.h [hunk 2]
```diff
@@ -89,6 +89,7 @@ struct i2c_dw_rom_config {
 	DEVICE_MMIO_ROM;
 	i2c_isr_cb_t	config_func;
 	uint32_t		bitrate;
+	const struct pinctrl_dev_config *pcfg;
```
Comment (gmarull) [inline drivers/i2c/i2c_dw.h:92]: field should be conditional

File: boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi [hunk 3]
```diff
@@ -15,4 +15,34 @@
...
+			bias-pull-up;
+			input-enable;
+			input-schmitt-enable;
+		};
+		group2 {
+			pinmux = <I2C0_SCL_P5>;
+			bias-pull-up;
+			input-enable;
+			input-schmitt-enable;
+		};
```
Comment (gmarull) [inline boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi:31]:
```suggestion
		group1 {
			pinmux = <I2C0_SDA_P4>, <I2C0_SCL_P5>;
			bias-pull-up;
			input-enable;
			input-schmitt-enable;
		};
```

note: I2C resistor value depends on bus speed, so using internal pull-ups is, in general, a bad idea.

File: boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi [hunk 4]
```diff
@@ -15,4 +15,34 @@
...
+		group2 {
+			pinmux = <I2C1_SCL_P15>;
+			bias-pull-up;
+			input-enable;
+			input-schmitt-enable;
+		};
```
Comment (gmarull) [inline boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi:46]:
```suggestion
		group1 {
			pinmux = <I2C1_SDA_P14>, <I2C1_SCL_P15>;
			bias-pull-up;
			input-enable;
			input-schmitt-enable;
		};
```

File: boards/arm/rpi_pico/rpi_pico.dts [hunk 5]
```diff
@@ -68,3 +68,17 @@
 &gpio0 {
 	status = "okay";
 };
+
+&i2c0 {
+	clock-frequency = <100000>;
```
Comment (gmarull) [inline boards/arm/rpi_pico/rpi_pico.dts:73]:
```suggestion
	clock-frequency = <I2C_BITRATE_STANDARD>;
```

ref include/dt-bindings/i2c/i2c.h

File: boards/arm/rpi_pico/rpi_pico.dts [hunk 6]
```diff
@@ -68,3 +68,17 @@
...
+	status = "okay";
+	pinctrl-0 = <&i2c0_default>;
+	pinctrl-names = "default";
+};
+
+&i2c1 {
+	clock-frequency = <100000>;
```
Comment (gmarull) [inline boards/arm/rpi_pico/rpi_pico.dts:80]:
```suggestion
	clock-frequency = <I2C_BITRATE_STANDARD>;
```

include/dt-bindings/i2c/i2c.h

**Review by gmarull — CHANGES_REQUESTED** [review]
Thanks for this PR, I have a few comments

File: drivers/i2c/i2c_dw.c (hunk 1 above)
Comment (andrei-edward-popa) [reply drivers/i2c/i2c_dw.c:849]: Yes, you are right, I create another commit with pinctrl support.

File: drivers/i2c/i2c_dw.h (hunk 2 above)
Comment (andrei-edward-popa) [reply drivers/i2c/i2c_dw.h:94]: I fixed this.

File: boards/arm/rpi_pico/rpi_pico.dts (hunk 5 above)
Comment (andrei-edward-popa) [reply boards/arm/rpi_pico/rpi_pico.dts]: I forgot about this macro, now it is fixed.

File: boards/arm/rpi_pico/rpi_pico.dts (hunk 6 above)
Comment (andrei-edward-popa) [reply boards/arm/rpi_pico/rpi_pico.dts]: I forgot about this macro, now it is fixed.

File: boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi (hunk 3 above)
Comment (andrei-edward-popa) [reply boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi]: Done!

File: boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi (hunk 4 above)
Comment (andrei-edward-popa) [reply boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi]: Done!

**Review by gmarull — APPROVED** [review]
LGTM, pinctrl and board parts. Note: please check my I2C pull-up comment.

**Comment by gmarull** [comment]
> > LGTM, pinctrl and board parts. Note: please check my I2C pull-up comment.
> 
> So, in this case, bias-pull-up needs to be removed?

I'm not 100% familiar with pico, so maybe its internal pull-ups are fine when working with I2C standard bitrate. You can check https://electronics.stackexchange.com/questions/1849/is-there-a-correct-resistance-value-for-i2c-pull-up-resistors

**Review by teburd — APPROVED** [review]
looks great to me, now I need to get a pico at some point

File: boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi [hunk 7]
```diff
@@ -15,4 +15,22 @@
 			input-enable;
 		};
 	};
+
+	i2c0_default: i2c0_default {
+		group1 {
+			pinmux = <I2C0_SDA_P4>, <I2C0_SCL_P5>;
+			bias-pull-up;
```
Comment (yonsch) [inline boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi:22]: I'm not sure pull up should be enabled by default, as it is common to have external pull ups (for example, if I2C is used for a sensor that already has pull ups)

File: boards/arm/rpi_pico/rpi_pico.dts [hunk 8]
```diff
@@ -68,3 +70,17 @@
 &gpio0 {
 	status = "okay";
 };
+
+&i2c0 {
+	clock-frequency = <I2C_BITRATE_STANDARD>;
+	status = "okay";
```
Comment (yonsch) [inline boards/arm/rpi_pico/rpi_pico.dts:76]: I don't think this should be enabled by default. If I'm not wrong, peripherals should be disabled by default, and enabled in an overlay file (with the console UART and GPIO being the exceptions). But input from others on this topic is welcomed.

File: boards/arm/rpi_pico/rpi_pico.yaml [hunk 9]
```diff
@@ -11,3 +11,4 @@ toolchain:
 supported:
   - uart
   - gpio
+  - i2c
```
Comment (yonsch) [inline boards/arm/rpi_pico/rpi_pico.yaml:14]: Please also add i2c to this table: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/rpi_pico/doc/index.rst#supported-features

**Review by yonsch — COMMENTED** [review]
A few comments/questions about the RPi part (I didn't look at the I2C driver itself).

File: boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi (hunk 7 above)
Comment (gmarull) [reply boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi:22]: agreed, ref https://github.com/zephyrproject-rtos/zephyr/pull/42513#issuecomment-1036411360

Comment (andrei-edward-popa) [reply boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi]: Yes, you are right, I removed it.

File: boards/arm/rpi_pico/rpi_pico.yaml (hunk 9 above)
Comment (andrei-edward-popa) [reply boards/arm/rpi_pico/rpi_pico.yaml:14]: I forgot about the docs, I2C is added now.

File: boards/arm/rpi_pico/rpi_pico.dts (hunk 8 above)
Comment (yonsch) [reply boards/arm/rpi_pico/rpi_pico.dts:76]: But the rp2040 has two i2c peripherals, one might enable i2c to use one and unintentionally enable the other. It happened to me once on an STM32 board, where UART stopped working because an i2c peripheral using the same pin was enabled by default

Comment (andrei-edward-popa) [reply boards/arm/rpi_pico/rpi_pico.dts:76]: So what do you suggest? To remove those ref nodes from dts or to make status disabled for both of them?

Comment (gmarull) [reply boards/arm/rpi_pico/rpi_pico.dts:76]: Applications can always override using an overlay, so if they want i2c1 and not i2c0, an overlay can be used:

```
&i2c0 {
    status = "disabled";
};
```

Comment (yonsch) [reply boards/arm/rpi_pico/rpi_pico.dts:76]: True, but In my opinion it's better to have to explicitly enable peripherals than having to explicitly disable them. The latter case makes the user responsible for making sure peripherals are not accidentally enabled, which assumes that the user is aware of this problem. Accidentally turning on a peripheral can increase power consumption and cause conflicts like I had.

Comment (andrei-edward-popa) [reply boards/arm/rpi_pico/rpi_pico.dts]: @yonsch I agree, I left the nodes in dts and I put status disabled on them.

Comment (andrei-edward-popa) [reply boards/arm/rpi_pico/rpi_pico.dts:76]: @yonsch Can you take a look at boards/arm/nucleo_f091rc/nucleo_f091rc.dts? It is similar with what I did for Pico's I2C.

Comment (yonsch) [reply boards/arm/rpi_pico/rpi_pico.dts:76]: Okay, if this is the policy I won't discuss it here, though I might continue this discussion somewhere else :) . Meanwhile, all my comments are resolved

Comment (teburd) [reply boards/arm/rpi_pico/rpi_pico.dts:76]: @yonsch sane defaults are certainly worth considering, but perhaps a broader topic. Can you please file an issue and mark it with dev-review perhaps if you want to discuss in a live setting. I generally agree that surprising defaults are a misfeature, but the way this is done here is similar to many other dts files in the tree today so certainly shouldn't be a blocker for this particular PR.

Comment (mbolivar-nordic) [reply boards/arm/rpi_pico/rpi_pico.dts:76]: I tend to agree with @yonsch. Here is our policy:

https://docs.zephyrproject.org/latest/guides/porting/board_porting.html#general-recommendations

In particular (emphasis mine):

> Unless explicitly recommended otherwise by this section, leave **peripherals and their drivers** disabled by default.

So @gmarull 's comment that the driver is enabled by default does not meet this requirement.

**Review by mbolivar-nordic — DISMISSED** [review]
Unless the I2C peripherals are being enabled in some way that the board porting guidelines allow, I don't think we should use other boards as an excuse for behavior against policy here.

**Comment by mbolivar-nordic** [comment]
On second thought, I think it's not worth blocking this PR over this, but I do want to make it clear that existing bad behavior is still bad behavior.

File: boards/arm/rpi_pico/rpi_pico.dts (hunk 8 above)
Comment (jfischer-no) [reply boards/arm/rpi_pico/rpi_pico.dts:76]: >It's ok to have status = "okay";. Board doesn't set CONFIG_I2C=y, so i2c driver won't be enabled unless applications do.

@gmarull 's comment is correct, decisive is `CONFIG_I2C` and reflects our general practice, it is also consistent with board porting guide.

Comment (gmarull) [reply boards/arm/rpi_pico/rpi_pico.dts:76]: If we let boards add `status = "disabled";`, it means that we'll need many overlays in samples that are board-specific, whereas `CONFIG_I2C=y` in `prj.conf` works for all boards that have an I2C enabled in DT. I think it's a bit confusing, but I can't think of a better way.

Comment (yonsch) [reply boards/arm/rpi_pico/rpi_pico.dts:76]: Opened an issue to continue this discussion: #42998

Comment (gmarull) [reply boards/arm/rpi_pico/rpi_pico.dts:76]: Note that this would likely not happen with a custom out-of-tree board. In those cases, you generally describe what you have/need, and the application tends to be specifically designed for that hardware. Upstream, we have lots of eval kits or demo boards that do not have a specific purpose. Or sometimes, boards are meant to be plugged into an expansion kit, where hardware really gets defined from a board perspective (see for example https://www.tindie.com/products/arturo182/rp2040-stamp/). For maintainability reasons, I'd avoid going the path of disabling peripherals and require sample overlays.
